Test Annotation Framework for Imported CPython test (#1973)
RustPython/RustPython: DEVELOPMENT.md

## Testing

RustPython uses three classes of tests. First, rust unit tests, second, our 
own tests written in python, and last, tests imported from CPython. 

Rust unit tests can be run with `cargo`:

```shell
$ cargo test --all
```

Our own collection of tests, called snippets, to test RustPython's functionality is located
in the `tests/snippets` directory and can be run using `pytest`:

```shell
$ cd tests
$ pytest -v
```

A set of annotated CPython tests are located in `Lib/tests`. Execute them by

```shell
$ cargo run --release -- -m test
```

### Creating tests
When importing tests from CPython, please follow the following rules.
1. Import complete files and document the version of the imported file and the major version this file relates to, e.g., 3.8.
2. In any case, keep the name of the original tests and test cases unmodified, such that it can always be traced to the original one from CPython.
3. Modify or substitute test cases, if required, to make them pass. We propose to

>  - modify test cases only when threre is only a minor deviation from the original behaviour. E.g., when the text of an error message does not match.

Reviewer: ```suggestion
 - modify test cases only when there is only a minor deviation from the original behaviour. E.g., when the text of an error message does not match.
```

>  - substitute test cases when there are severe deviations from the original is required. E.g., when it is necessary to introduce new variables or change the logic of the test or split a test case in several test cases.

Reviewer: ```suggestion
 - substitute test cases when severe deviations from the original is required. E.g., when it is necessary to introduce new variables or change the logic of the test or split a test case in several test cases.
```

4. Add further test cases.
5. Annotate each test case with:
 - Original and unmodified test case that is supposed to be passed: no annotation is required, but when you want to be more verbose you can use @rpt.imported.original
 - Original and unmodified test case that is supposed to be failed: @rpt.imported.fail
 - Original and unmodified test case that is not executing, causes panic, breaks RustPython, etc.: @rpt.imported.skip
 - Original but modified test case: @rpt.imported.modified

>  - Substituted test cases: @prt.imported.substituted (Note: this leads to skipping this test as it is considered as distirbung, by setting the parameter `run=True` it is executed but expected to fail.)

Reviewer: ```suggestion
 - Substituted test cases: @prt.imported.substituted (Note: this leads to skipping this test as it is considered as distirbung, by setting the parameter `run=True` it is executed but expected to fail.)
```
Not sure what "distirbung" is meant to say

>  - New test cases that substitue an original test case: @rpt.ours.subst and refer to the substituted original test cases.

Reviewer: ```suggestion
 - New test cases that substitute for an original test case: @rpt.ours.subst and refer to the substituted original test cases.
```

Understanding a new codebase takes time. Here's a brief view of the
repository's structure:

> .

Reviewer 2: ```suggestion
```
